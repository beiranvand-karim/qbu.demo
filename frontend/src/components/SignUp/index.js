import React, {Component} from 'react'
import {Button, Input, Label, FormGroup, Form} from "reactstrap"
import config from "../../config"
import PageCenter from "../../StyledComponents/PageCenter"

const empty = '';

export class SignUp extends Component {

   constructor(props) {
      super(props);
      this.signUpFormSubmission = this.signUpFormSubmission.bind(this);
      this.change = this.change.bind(this)
   }

   state = {
      firstName: empty,
      lastName: empty,
      username: empty,
      password: empty,
      email: empty,
      phone: empty,
      data: null,
      loading: false,
      error: null
   };

   signUpServerCall(user) {
      this.setState({loading: true});
      fetch(`${config.server}/user/signUp`, {
         method: "POST",
         headers: {
            "Content-Type": "application/json"
         },
         body: JSON.stringify(user)
      })
         .then(resp => resp.json())
         .then(data => {
            this.setState({loading: false});
            this.setState({data})
         })
         .catch(error => {
            this.setState({loading: false});
            this.setState({error})
         })
   }

   signUpFormSubmission(e) {
      e.preventDefault();
      const {firstName, lastName, username, password, email, phone} = this.state;
      const user = {firstName, lastName, username, password, email, phone};
      this.signUpServerCall(user)
   };

   change(e) {
      this.setState({
         [e.target.name]: e.target.value
      })
   }

   render() {

      if (this.state.data) {
         return <h1>user created</h1>
      }

      if (this.state.error) {
         return <h4>error: {this.state.error}</h4>
      }

      if (this.state.loading) {
         return <h3>loading ...</h3>
      }

      return <>
         <PageCenter>
            <Form onSubmit={this.signUpFormSubmission}>
               <FormGroup>
                  <Label for="firstName">first name</Label>
                  <Input value={this.state.firstName}
                         onChange={this.change}
                         type="text"
                         name="firstName"
                         id="firstName"
                         placeholder="first name"/>
               </FormGroup>
               <FormGroup>
                  <Label for="lastName">last name</Label>
                  <Input value={this.state.lastName}
                         onChange={this.change}
                         type="text" name="lastName" id="lastName"
                         placeholder="last name"/>
               </FormGroup>
               <FormGroup>
                  <Label for="username">username</Label>
                  <Input value={this.state.username}
                         onChange={this.change}
                         type="text" name="username" id="username"
                         placeholder="username"/>
               </FormGroup>
               <FormGroup>
                  <Label for="password">password</Label>
                  <Input value={this.state.password}
                         onChange={this.change}
                         type="password" name="password" id="password"
                         placeholder="password"/>
               </FormGroup>
               <FormGroup>
                  <Label for="email">email</Label>
                  <Input value={this.state.email}
                         onChange={this.change}
                         type="email" name="email" id="email"
                         placeholder="email"/>
               </FormGroup>
               <FormGroup>
                  <Label for="phone">phone</Label>
                  <Input value={this.state.phone}
                         onChange={this.change}
                         type="text" name="phone" id="phone"
                         placeholder="phone"/>
               </FormGroup>
               <Button>Submit</Button>
            </Form>
         </PageCenter>
      </>
   }

};

export default SignUp