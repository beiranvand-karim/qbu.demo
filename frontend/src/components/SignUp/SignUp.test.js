/* global describe, it, expect, jest */
import React from 'react'
import SignUp from './'
import {shallow} from 'enzyme'
import renderer from 'react-test-renderer'
import {MemoryRouter} from "react-router-dom";

describe('<SignUp />', () => {

   const user = {
      "id": 14,
      "username": "megan",
      "password": "$2a$12$GMpxEokq3TsYeKMFd.gd8urSAYFKEUUOKTCc0Hj5Z31BAc/7Ly5li",
      "firstName": "megan",
      "lastName": "fox",
      "email": "fox.megan@gmail.com",
      "phone": null,
      "enabled": true,
      "authorities": [
         {
            "authority": "ROLE_USER"
         }
      ],
      "accountNonExpired": true,
      "accountNonLocked": true,
      "credentialsNonExpired": true
   };

   it('should render without crashing', () => {
      shallow(<SignUp/>)
   });

   it('should sign up a user', () => {
      window.fetch = jest.fn().mockImplementation(() => {
            return new Promise((resolve) => {
               resolve({
                  status: 200,
                  json: () => {
                     return new Promise((resolve) => {
                        resolve({
                           user
                        })
                     })
                  }
               })
            })
         }
      );
      const wrapper = shallow(<SignUp/>);
      wrapper.find('[name="firstName"]').simulate('change', {target: {name: 'firstName', value: 'megan'}});
      wrapper.find('[name="lastName"]').simulate('change', {target: {name: 'lastName', value: 'fox'}});
      wrapper.find('[name="username"]').simulate('change', {target: {name: 'username', value: 'megan'}});
      wrapper.find('[name="password"]').simulate('change', {target: {name: 'password', value: 'megan852654'}});
      wrapper.find('[name="email"]').simulate('change', {target: {name: 'email', value: 'fox.megan@gmail.com'}});
      wrapper.find('[name="phone"]').simulate('change', {target: {name: 'phone', value: '09906042587'}});
      wrapper.find('Form').simulate('submit', {preventDefault(){}});

      expect(wrapper.state().loading).toEqual(true);

      setImmediate(() => {
         expect(wrapper.state().loading).toEqual(false);
         expect(wrapper.state().data).toEqual({user});
      })

   });

   it('should fail to sign up a user', () => {

      const error = 'sign up failed.';

      window.fetch = jest.fn().mockImplementation(() => {
            return new Promise((resolve,reject) => {
               reject({
                  error
               })
            })
         }
      );

      const wrapper = shallow(<SignUp/>);
      wrapper.find('[name="firstName"]').simulate('change', {target: {name: 'firstName', value: 'megan'}});
      wrapper.find('[name="lastName"]').simulate('change', {target: {name: 'lastName', value: 'fox'}});
      wrapper.find('[name="username"]').simulate('change', {target: {name: 'username', value: 'megan'}});
      wrapper.find('[name="password"]').simulate('change', {target: {name: 'password', value: 'megan852654'}});
      wrapper.find('[name="email"]').simulate('change', {target: {name: 'email', value: 'fox.megan@gmail.com'}});
      wrapper.find('[name="phone"]').simulate('change', {target: {name: 'phone', value: '09906042587'}});
      wrapper.find('Form').simulate('submit', {preventDefault(){}});

      expect(wrapper.state().loading).toEqual(true);

      setImmediate(() => {
         expect(wrapper.state().loading).toEqual(false);
         expect(wrapper.state().error).toEqual({error});
      })

   });

   it('should matches the snapshot', () => {
      const tree = renderer.create(<MemoryRouter><SignUp /></MemoryRouter>).toJSON();
      expect(tree).toMatchSnapshot()
   });

});