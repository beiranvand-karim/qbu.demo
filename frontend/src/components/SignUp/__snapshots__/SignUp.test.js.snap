// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> should matches the snapshot 1`] = `
<div
  className="sc-bdVaJa gtqnvx"
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      className="position-relative form-group"
    >
      <label
        className=""
        htmlFor="firstName"
      >
        first name
      </label>
      <input
        className="form-control"
        id="firstName"
        name="firstName"
        onChange={[Function]}
        placeholder="first name"
        type="text"
        value=""
      />
    </div>
    <div
      className="position-relative form-group"
    >
      <label
        className=""
        htmlFor="lastName"
      >
        last name
      </label>
      <input
        className="form-control"
        id="lastName"
        name="lastName"
        onChange={[Function]}
        placeholder="last name"
        type="text"
        value=""
      />
    </div>
    <div
      className="position-relative form-group"
    >
      <label
        className=""
        htmlFor="username"
      >
        username
      </label>
      <input
        className="form-control"
        id="username"
        name="username"
        onChange={[Function]}
        placeholder="username"
        type="text"
        value=""
      />
    </div>
    <div
      className="position-relative form-group"
    >
      <label
        className=""
        htmlFor="password"
      >
        password
      </label>
      <input
        className="form-control"
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </div>
    <div
      className="position-relative form-group"
    >
      <label
        className=""
        htmlFor="email"
      >
        email
      </label>
      <input
        className="form-control"
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="email"
        value=""
      />
    </div>
    <div
      className="position-relative form-group"
    >
      <label
        className=""
        htmlFor="phone"
      >
        phone
      </label>
      <input
        className="form-control"
        id="phone"
        name="phone"
        onChange={[Function]}
        placeholder="phone"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label={null}
      className="btn btn-secondary"
      onClick={[Function]}
    >
      Submit
    </button>
  </form>
</div>
`;
